### PR TITLE
HNSW refactor 3 - [MOD-5302]

### PR DESCRIPTION
**Pull request by GuyAv46**
**Describe the changes in the pull request**

3rd time is the charm. Refactor HNSW for better memory management using vector (and metadata) blocks.
It also improves reliability using CPP structs instead of calculating offsets from the entire index allocation monolith.

Currently not supporting old serialization versions, may pr may not happen in another PR.

**Main objects this PR modified**
1. HNSW internal implementation (data structures, resize logic)
2. initial capacity is now rounded up to the next block size, to simplify resizing of different data structures (both in BF and HNSW)
3. tuned `EstimateElementSize` for blocks growth

**Mark if applicable**

- [ ] This PR introduces API changes
- [X] This PR introduces serialization changes

### CODE REVIEW

File: src/VecSim/algorithms/hnsw/hnsw.h [hunk 1]
```diff
@@ -64,6 +65,46 @@ struct AddVectorCtx {
...
+            }
+            for (size_t i = 0; i < maxLevel; i++) {
+                new ((char *)others + i * high_level_size) level_data(allocator);
+            }
+        }
+    }
+};
+
+#pragma pack() // restore default packing
+
```
Comment (alonre24) [inline src/VecSim/algorithms/hnsw/hnsw.h:107]: Have a clear separation here... like #### or something

File: src/VecSim/algorithms/hnsw/hnsw.h [hunk 2]
```diff
@@ -1777,38 +2008,51 @@ void HNSWIndex<DataType, DistType>::removeVectorInPlace(const idType element_int
...
-    memset(data_level0_memory_ + state.newElementId * size_data_per_element_ + offsetLevel0_, 0,
-           size_data_per_element_);
-    // We mark id as in process *before* we set it in the label lookup, otherwise we might check
-    // that the label exist with safeCheckIfLabelExistsInIndex and see that IN_PROCESS flag is
-    // clear.
-    markInProcess(state.newElementId);
-    setVectorId(label, state.newElementId);
-    element_levels_[state.newElementId] = state.elementMaxLevel;
+    // Create the new element's metadata.
+    char tmpData[this->element_graph_data_size_];
```
Comment (alonre24) [inline src/VecSim/algorithms/hnsw/hnsw.h:2022]: Explain why you are allocating this amount of bytes

**Review by alonre24 — COMMENTED** [review]
BRAVO

File: src/VecSim/algorithms/brute_force/brute_force.h [hunk 3]
```diff
@@ -129,41 +151,35 @@ BruteForceIndex<DataType, DistType>::BruteForceIndex(
...
-    // Get the last vectors block to store the vector in (we assume that it's not full yet).
-    VectorBlock *vectorBlock = this->vectorBlocks.back();
-    assert(vectorBlock == getVectorVectorBlock(id));
-
-    // add vector data to vectorBlock
-    vectorBlock->addVector(vector_data);
+    // If `vectorBlocks` vector is empty or the last vector block is full, create a new block.
+    if (indexSize() > indexCapacity()) {
+        growByBlock();
+    } else if (id % this->blockSize == 0) {
```
[the remark this answers is not in the document]
Comment (GuyAv46) [reply src/VecSim/algorithms/brute_force/brute_force.h:173]: Right. Improved the comment

File: src/VecSim/utils/serializer.cpp [hunk 4]
```diff
@@ -22,10 +22,14 @@ Serializer::EncodingVersion Serializer::ReadVersion(std::ifstream &input) {
     // The version number is the first field that is serialized.
     EncodingVersion version = EncodingVersion_INVALID;
     readBinaryPOD(input, version);
-    // Only V1 and V2 are supported
-    if (version <= 0 || version >= EncodingVersion_INVALID) {
+    if (version <= EncodingVersion_DEPRECATED) {
```
[the remark this answers is not in the document]
Comment (GuyAv46) [reply src/VecSim/utils/serializer.cpp:25]: For now. TBD

File: tests/benchmark/bm_datasets.py [hunk 5]
```diff
@@ -97,12 +95,14 @@ def measure_recall_per_second(hnsw_index, dataset, is_multi, data_type, num_quer
     correct = 0
     bf_total_time = 0
     hnsw_total_time = 0
+    print(os.getpid())
```
[the remark this answers is not in the document]
Comment (GuyAv46) [reply tests/benchmark/bm_datasets.py:98]: Leftovers from benchmarking. Will remove

File: src/python_bindings/bindings.cpp [hunk 6]
```diff
@@ -545,10 +545,8 @@ PYBIND11_MODULE(VecSim, m) {
     py::class_<PyHNSWLibIndex, PyVecSimIndex>(m, "HNSWIndex")
         .def(py::init([](const HNSWParams &params) { return new PyHNSWLibIndex(params); }),
              py::arg("params"))
-        .def(py::init([](const std::string &location, const HNSWParams *params) {
-                 return new PyHNSWLibIndex(location, params);
-             }),
-             py::arg("location"), py::arg("params") = nullptr)
+        .def(py::init([](const std::string &location) { return new PyHNSWLibIndex(location); }),
```
[the remark this answers is not in the document]
Comment (GuyAv46) [reply src/python_bindings/bindings.cpp:548]: Right

File: src/VecSim/algorithms/hnsw/hnsw.h [hunk 7]
```diff
@@ -1732,9 +1728,9 @@ void HNSWIndex<DataType, DistType>::repairNodeConnections(idType node_id, size_t
                 continue;
             }
             neighbors_candidates_set[node_meta.links[j]] = true;
-            neighbors_candidates.emplace(this->dist_func(getDataByInternalId(node_id),
-                                                         getDataByInternalId(node_meta.links[j]),
-                                                         this->dim),
+            neighbors_candidates.emplace(this->distFunc(getDataByInternalId(node_id),
```
Comment (GuyAv46) [inline src/VecSim/algorithms/hnsw/hnsw.h:1731]: TODO: keep `getDataByInternalId(node_id)` address once at the beggining

File: src/VecSim/algorithms/hnsw/hnsw_serializer.h [hunk 8]
```diff
@@ -117,33 +113,33 @@ HNSWIndexMetaData HNSWIndex<DataType, DistType>::checkIntegrity() const {
 template <typename DataType, typename DistType>
 void HNSWIndex<DataType, DistType>::restoreIndexFields(std::ifstream &input) {
     // Restore index build parameters
-    readBinaryPOD(input, this->M_);
-    readBinaryPOD(input, this->maxM_);
-    readBinaryPOD(input, this->maxM0_);
-    readBinaryPOD(input, this->ef_construction_);
+    readBinaryPOD(input, this->M);
+    readBinaryPOD(input, this->M);
```
Comment (GuyAv46) [inline src/VecSim/algorithms/hnsw/hnsw_serializer.h:117]: TODO: remove duplication

File: src/VecSim/algorithms/hnsw/hnsw_serializer.h [hunk 9]
```diff
@@ -238,85 +234,85 @@ void HNSWIndex<DataType, DistType>::saveIndexFields(std::ofstream &output) const
...
+    writeBinaryPOD(output, this->maxElements); // This will be used to restore the index initial
+                                               // capacity
 
     // Save index build parameters
-    writeBinaryPOD(output, this->M_);
-    writeBinaryPOD(output, this->maxM_);
-    writeBinaryPOD(output, this->maxM0_);
-    writeBinaryPOD(output, this->ef_construction_);
+    writeBinaryPOD(output, this->M);
+    writeBinaryPOD(output, this->M);
```
Comment (GuyAv46) [inline src/VecSim/algorithms/hnsw/hnsw_serializer.h:242]: TODO: remove duplication

**Review by DvirDukhan — APPROVED** [review]

**Pull request merged**